### PR TITLE
Use gradients in sequential CES experiment

### PR DESCRIPTION
**Pull request by ae-foster**

### CODE REVIEW

File: pyro/contrib/oed/differentiable_eig.py [hunk 1]
```diff
@@ -30,14 +27,21 @@ def loss_fn(design, num_particles, control_variate=0., **kwargs):
...
 
         # Calculate the score parts
         trace.compute_score_parts()
         prescore_function = sum(trace.nodes[l]["score_parts"][1] for l in observation_labels)
+        logging.debug('prescore {}'.format(is_bad(prescore_function)))
         terms += (terms.detach() - control_variate) * prescore_function
 
-        return _safe_mean_terms(terms)
+        result = _safe_mean_terms(terms)
+        logging.debug("loss {}".format(result[0]))
```
Comment (ae-foster) [inline pyro/contrib/oed/differentiable_eig.py]: these logging statements can be removed

File: examples/contrib/oed/ces.py [hunk 2]
```diff
@@ -105,29 +184,33 @@ def main(num_steps, num_parallel, experiment_name, typs, seed, lengthscale):
             pyro.set_rng_seed(seed)
         marginal_mu_init, marginal_log_sigma_init = 0., 6.
         oed_n_samples, oed_n_steps, oed_final_n_samples, oed_lr = 10, 1250, 2000, [0.1, 0.01, 0.001]
-        elbo_n_samples, elbo_n_steps, elbo_lr = 10, 1000, 0.04
+        elbo_n_samples, elbo_n_steps, elbo_lr = 10, 6000, 0.01
```
Comment (ae-foster) [inline examples/contrib/oed/ces.py]: may be worth changing this back

File: examples/contrib/oed/ces_gradients.py [hunk 3]
```diff
@@ -11,53 +11,67 @@
...
 
 
+def proposal(design):
+
+        batch_shape = design.shape[:-2]
+        with ExitStack() as stack:
+            for plate in iter_plates_to_shape(batch_shape):
+                stack.enter_context(plate)
+            emission_dist = dist.CensoredSigmoidNormal(torch.tensor([0.]), torch.tensor([50.]), 1 - epsilon, epsilon).to_event(1)
+            pyro.sample("y", emission_dist)
```
Comment (ae-foster) [inline examples/contrib/oed/ces_gradients.py]: probably no longer needed

File: examples/contrib/oed/results_parser_ces.py [hunk 4]
```diff
@@ -21,18 +21,20 @@
...
-    centre = array.mean(1)
-    se = array.std(1)/np.sqrt(array.shape[1])
-    upper, lower = centre + se, centre - se
-    return lower, centre, upper
-    # return np.percentile(array, 75, axis=1), np.percentile(array, 50, axis=1), np.percentile(array, 25, axis=1)
+    # centre = array.mean(1)
+    # se = array.std(1)/np.sqrt(array.shape[1])
+    # upper, lower = centre + se, centre - se
+    # return lower, centre, upper
+    return np.percentile(array, 75, axis=1), np.percentile(array, 50, axis=1), np.percentile(array, 25, axis=1)
```
Comment (ae-foster) [inline examples/contrib/oed/results_parser_ces.py]: Change this back!

File: examples/contrib/oed/ces.py [hunk 5]
```diff
@@ -1,58 +1,84 @@
...
             U2rho = (rmv(d2.pow(rho.unsqueeze(-1)), alpha)).pow(1./rho)
             mean = slope * (U1rho - U2rho)
-            # print('latent samples:', rho.mean(), alpha.mean(), slope.mean(), slope.median())
-            # print('mean', mean.mean(), mean.std(), mean.min(), mean.max())
+            print('slope', slope.min().item(), slope.max().item())
+            print('u diff', (U1rho - U2rho).min().item(), (U1rho - U2rho).max().item())
+            logging.debug('latent samples: rho {} alpha {} slope mean {} slope median {}'.format(
+                rho.mean().item(), alpha.mean().item(), slope.mean().item(), slope.median().item()))
+            print('mean: mean {} sd {} min {} max {}'.format(
+                mean.mean().item(), mean.std().item(), mean.min().item(), mean.max().item()))
```
Comment (ae-foster) [inline examples/contrib/oed/ces.py:58]: clean up the debugging output

File: examples/contrib/oed/ces.py [hunk 6]
```diff
@@ -86,17 +110,121 @@ def guide(design, observation_labels, target_labels):
...
+        slope_sigma = self.prior_slope_sigma * (1e-6 + self.softplus(final[..., 6])) 
+        #rho_concentration = 1e-6 + self.softplus(self.rho_concentration(x))
+        #alpha_concentration = 1e-6 + self.softplus(self.alpha_concentration(x))
+        #slope_mu = self.slope_mu(x).squeeze(-1)
+        #slope_sigma = 1e-6 + self.softplus(self.slope_sigma(x)).squeeze(-1)
+        print('slope distribution', slope_sigma.min(), slope_sigma.max(), slope_mu.min(), slope_mu.max())
+        rho_concentration.register_hook(lambda x: print('rho grad', x.max(), x.min()))
+        alpha_concentration.register_hook(lambda x: print('alpha grad', x.max(), x.min()))
+        slope_mu.register_hook(lambda x: print('slope mu grad', x.max(), x.min()))
+        #slope_sigma.register_hook(lambda x: print('slope sigma grad', x.max(), x.min()))
```
Comment (ae-foster) [inline examples/contrib/oed/ces.py]: clean this up

File: examples/contrib/oed/ces_gradients.py [hunk 7]
```diff
@@ -70,11 +85,10 @@ def forward(self, y_dict, design_prototype, observation_labels, target_labels):
...
         x = self.softplus(self.linear1(s))
         x = self.softplus(self.linear2(x))
-        rho0 = self.softplus(self.rho0(x)).squeeze(-1)
-        rho1 = self.softplus(self.rho1(x)).squeeze(-1)
-        alpha_concentration = self.softplus(self.alpha_concentration(x))
+        rho_concentration = 1e-6 + self.softplus(self.rho_concentration(x))
+        alpha_concentration = 1e-6 + self.softplus(self.alpha_concentration(x))
         slope_mu = self.slope_mu(x).squeeze(-1)
-        slope_sigma = self.softplus(self.slope_sigma(x)).squeeze(-1)
+        slope_sigma = 1e-6 + self.softplus(self.slope_sigma(x)).squeeze(-1)
```
Comment (ae-foster) [inline examples/contrib/oed/ces_gradients.py]: match neural architectures between ces and ces_gradients. Maybe even import into ces to save space

File: examples/contrib/oed/results_parser_ces.py [hunk 8]
```diff
@@ -72,8 +74,14 @@ def main(fnames, findices, plot):
             try:
                 while True:
                     results = pickle.load(results_file)
+                    print(results['design_time'])
```
Comment (ae-foster) [inline examples/contrib/oed/results_parser_ces.py]: remove

File: pyro/contrib/oed/differentiable_eig.py [hunk 9]
```diff
@@ -60,62 +59,129 @@ def differentiable_nce_eig(model, design, observation_labels, target_labels=None
...
     prescore_function = sum(trace.nodes[l]["score_parts"][1] for l in observation_labels)
-    terms += (terms.detach() - control_variate) * prescore_function
+    grad_terms = (terms.detach() - control_variate) * prescore_function
 
-    return _safe_mean_terms(terms)
+    surrogate_loss = _safe_mean_terms(grad_terms)[0]
+    return (surrogate_loss, nce_part)
+
+
+def differentiable_nce_proposal_eig(model, design, observation_labels, target_labels,
```
Comment (ae-foster) [inline pyro/contrib/oed/differentiable_eig.py:78]: comment out, or leave deprecated for now

File: pyro/contrib/oed/differentiable_eig.py [hunk 10]
```diff
@@ -60,62 +59,129 @@ def differentiable_nce_eig(model, design, observation_labels, target_labels=None
...
+
+    # Calculate the score parts
+    grad_terms = (terms.detach() - control_variate) * importance_weights
+
+    surrogate_loss = _safe_mean_terms(grad_terms)[0]
+    return (surrogate_loss, nce_part)
+
+
+def desc(t):
+    return "max {} min {} median {}".format(t.max().item(), t.min().item(), t.median().item())
```
Comment (ae-foster) [inline pyro/contrib/oed/differentiable_eig.py]: delete

File: pyro/contrib/oed/differentiable_eig.py [hunk 11]
```diff
@@ -60,62 +59,129 @@ def differentiable_nce_eig(model, design, observation_labels, target_labels=None
...
 
 
 def _differentiable_ace_eig_loss(model, guide, M, observation_labels, target_labels):
```
Comment (ae-foster) [inline pyro/contrib/oed/differentiable_eig.py:115]: remove print statements- surrogate loss may need to change

File: pyro/contrib/oed/eig.py [hunk 12]
```diff
@@ -631,6 +632,7 @@ def opt_eig_ape_loss(design, loss_fn, num_samples, num_steps, optim, return_hist
     params = None
     history = []
     for step in range(num_steps):
+        print(step)
```
Comment (ae-foster) [inline pyro/contrib/oed/eig.py]: delete

File: pyro/contrib/oed/eig.py [hunk 13]
```diff
@@ -639,10 +641,19 @@ def opt_eig_ape_loss(design, loss_fn, num_samples, num_steps, optim, return_hist
                      for site in param_capture.trace.nodes.values())
         if torch.isnan(agg_loss):
             raise ArithmeticError("Encountered NaN loss in opt_eig_ape_loss")
+        print('agg_loss', agg_loss)
         agg_loss.backward(retain_graph=True)
+        print('after backward')
```
Comment (ae-foster) [inline pyro/contrib/oed/eig.py]: delete

File: pyro/distributions/censored_sigmoid_normal.py [hunk 14]
```diff
@@ -87,7 +87,7 @@ def log_prob(self, value):
         lower_cdf = lower_cdf.log()
         lower_cdf[mask_lower] = asymptotic_lower[mask_lower]
         if is_bad(upper_cdf):
-            raise ArithmeticError("NaN in upper cdf {}".format(upper_cdf))
+            raise ArithmeticError("NaN in upper cdf")
```
Comment (ae-foster) [inline pyro/distributions/censored_sigmoid_normal.py:90]: potentially print the median, min, max

File: examples/contrib/oed/results_parser_ces.py [hunk 15]
```diff
@@ -101,9 +108,9 @@ def main(fnames, findices, plot):
         for statistic in ["Entropy", "rho_rmse", "alpha_rmse", "slope_rmse"]:
             plt.figure(figsize=(5, 5))
             for i, k in enumerate(reformed[statistic]):
-                e = reformed[statistic][k].squeeze()
+                e = reformed[statistic][k].squeeze()[1:]
                 lower, centre, upper = upper_lower(e)
-                x = np.arange(1, e.shape[0]+1)
+                x = np.arange(2, e.shape[0]+2)
```
Comment (ae-foster) [inline examples/contrib/oed/results_parser_ces.py:113]: do we want this on branch?
